### PR TITLE
Feature/csec 12868 upgrade operator repo

### PR DESCRIPTION
**Pull request by tbronsonCoveo**
Ignore the ridiculous number of changes, 90% of those are poetry lock files that are managed automatically.
This PR reflects the changes of https://github.com/coveooss/aws_auth_eks_crd/pull/8 in the operator.

1- Changed the operator to update rolearn mappings (previously only supported userarn) in the corresponding  mapRoles entry of the aws-auth kube-system configmap.
2- Added poetry to install dependencies and linters with pre-commit (https://pre-commit.com/) to clean the code.
3- Added tests to cover the functionalities that were added.
4-Modified Dockerfile to reflect the directory changes. Added non root user to execute the operator.

Git marked the whole iam_mapping.py as new. Although I didn't rewrite the whole module, I did fix some bugs so I would recommend reviewing the whole iam_mapping file. That's where all the logic changes were made to process the mapUsers and mapRoles mappings together.

### CODE REVIEW

File: src/kubernetes_operator/iam_mapping.py
Comment (sbeginCoveo) [inline src/kubernetes_operator/iam_mapping.py:205]: we could use a python `filter` here
```
return filter(lambda existing_identity: existing_identity["username"] != identity["username"], identity_list)
```

File: pyproject.toml
Comment (sbeginCoveo) [inline pyproject.toml:15]: nice upgrade :+1:

File: src/kubernetes_operator/iam_mapping.py
Comment (tbronsonCoveo) [reply src/kubernetes_operator/iam_mapping.py:103]: Not a fan either. But I left it as is. I believe the diff is from formatting. 

`Define a liveness HTTP request: Any code greater than or equal to 200 and less than 400 indicates success. Any other code indicates failure.` According to https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-liveness-http-request. 

I am not sure how kopf handles the result response code. The doc doesn't seem to mention anything on the subject https://kopf.readthedocs.io/en/latest/probing/. Raising was probably assumed to be a guaranteed way to force an error code when it was originally coded. Unless you guys are aware of another way to trigger this, I will leave the exception and apply your comment.

File: src/kubernetes_operator/iam_mapping.py
Comment (tbronsonCoveo) [reply src/kubernetes_operator/iam_mapping.py:1]: Heads up, the diff is influenced by the formatting. But feel free to comment on any of the code and I'll go over it even if it wasn't a change. Might as well improve it.

File: src/kubernetes_operator/iam_mapping.py
Comment (tbronsonCoveo) [reply src/kubernetes_operator/iam_mapping.py:189]: Although it makes it cleaner, I just realized it would break updates. This method does more than add the identity. It also updates the existing one with the one passed in the params. I will add it to the doc and write a test for that case, didn't think of that.

File: src/kubernetes_operator/iam_mapping.py
Comment (tbronsonCoveo) [reply src/kubernetes_operator/iam_mapping.py:205]: I prefer this shorter version but it will not signal any warning if the identity is not found. 
Not sure when that case would happen but it would definitely be a warning you would want in the logs if it does happen. What are your thoughts on that?

File: .gitignore
Comment (tbronsonCoveo) [reply .gitignore:5]: Not sure what you guys use. Could you link a gitignore you usually use in the infra projects?

File: src/kubernetes_operator/iam_mapping.py
Comment (sbeginCoveo) [reply src/kubernetes_operator/iam_mapping.py:207]: I think youre right, we could compare the length of both lists before returning, something like
```
new_list = filter(...)
if len(new_list) == len(original_list):
    log a warning
return new_list
```
although at this point, im being nit picky :P I tend to prefer having single return pathways, but you have already put more time in this repo than me

File: src/kubernetes_operator/iam_mapping.py
Comment (sbeginCoveo) [reply src/kubernetes_operator/iam_mapping.py]: I mean, the exception is fine. Godspeed to raising exceptions, they are really useful.
It was just having it as the last/default return pathway of the function which felt counter-intuitive. And i also realise it was like that before your changes

File: .gitignore
Comment (sbeginCoveo) [reply .gitignore:5]: its fine, we usually add patterns only when necessary, [ex](https://github.com/coveo/infra-script/blob/master/.gitignore)
having too much doesn't hurt anyone :)

File: .gitignore
Comment (JHeilCoveo) [inline .gitignore:6]: Please re-add Visual Studio Code.

File: .pre-commit-config.yaml
Comment (JHeilCoveo) [inline .pre-commit-config.yaml:23]: Pre-commit hook is fine but I prefer if the settings for black, mypy and friends are committed as settings in the repository so my tools do fight with this pre-commit.

File: src/kubernetes_operator/iam_mapping.py
Comment (JHeilCoveo) [inline src/kubernetes_operator/iam_mapping.py:44]: Hard to read condition IMHO, would probably be better to write it like this
```suggestion
    if not diff:
        return
```

Then de-indent the rest

File: src/kubernetes_operator/iam_mapping.py
Comment (JHeilCoveo) [inline src/kubernetes_operator/iam_mapping.py:108]:
```suggestion
    crd_file_path = get_project_root() / "kubernetes" / "iamidentitymappings.yaml"
```

File: src/kubernetes_operator/iam_mapping.py
Comment (JHeilCoveo) [inline src/kubernetes_operator/iam_mapping.py:154]: I don't know, shouldn't we let this operator crash instead of catching it and printing a warning ? We could erase all mappings with this.

File: src/kubernetes_operator/iam_mapping.py
Comment (JHeilCoveo) [reply src/kubernetes_operator/iam_mapping.py:207]: @sbeginCoveo FYI if you filter(...) you will return an iterable and not a list which most of time works but could break in some situations.

File: src/kubernetes_operator/iam_mapping.py
Comment (JHeilCoveo) [inline src/kubernetes_operator/iam_mapping.py:210]: Those 2 functions would probably have triggered me to rewrite this using Dataclasses and Sets instead of manually iterating over the list and updating it. But hey who Am I to judge ? :stuck_out_tongue:

File: src/kubernetes_operator/iam_mapping.py
Comment (JHeilCoveo) [inline src/kubernetes_operator/iam_mapping.py:221]: I don't like this but I don't see any better solution :(

File: src/kubernetes_operator/iam_mapping.py
Comment (tbronsonCoveo) [reply src/kubernetes_operator/iam_mapping.py:154]: Forgot to change it. Will do.

File: src/kubernetes_operator/iam_mapping.py
Comment (tbronsonCoveo) [reply src/kubernetes_operator/iam_mapping.py:221]: Yeah unless we want to stick the yaml in the src folder, I never found a clean way to do this in python.

File: src/kubernetes_operator/iam_mapping.py
Comment (tbronsonCoveo) [reply src/kubernetes_operator/iam_mapping.py:210]: Definitely agree, although I did say we could improve it in this PR, it wasn't the main goal. So I am thinking of taking a pass on this one. There aren't any tests so if I rewrite I would also have to test. I think I'll leave it to you guys later on if you need it ;)

File: src/kubernetes_operator/iam_mapping.py
Comment (sbeginCoveo) [reply src/kubernetes_operator/iam_mapping.py:207]: of course, it has to pass through `list(...)` as well :)

File: .pre-commit-config.yaml
Comment (tbronsonCoveo) [reply .pre-commit-config.yaml:23]: I transfered black but unfortunately pycodestyle only supports tox.ini or setup.cfg files... so I added the tox.ini file with only pycodestyle settings to avoid problems with your tools, although it isn't very clean.

**Review by JHeilCoveo — APPROVED** [review]
LGTM :+1:

File: src/kubernetes_operator/iam_mapping.py
Comment (JHeilCoveo) [reply src/kubernetes_operator/iam_mapping.py:223]: Maybe something with https://setuptools.readthedocs.io/en/latest/pkg_resources.html#resourcemanager-api ?
Anyway not blocking.

File: src/kubernetes_operator/iam_mapping.py
Comment (JPLachance) [inline src/kubernetes_operator/iam_mapping.py:75]: I just did a small test:
- Apply a aws-auth CM in the kube-system Namespace with stuff in it (stuff you won't find in a IAMIdentityMapping), like ours here https://github.com/coveo/infra-kubernetes/blob/7cde0cc290558dbd36eac3ff22d2a87a23466010/infrastructure/terraform/coveo-eks/aws_auth_cm.tf
- Boot this Operator
- Watch the operator remove everything

Booting this Operator in a live Coveo EKS cluster will effectively brick the cluster 💥 

We need to take into account this aws-auth CM will not be empty, or we need to create one IAMIdentityMapping in the infra-kubernetes repo for each entry in our detault CM.

I'll let your decide with the Infra team.

**Review by JPLachance — CHANGES_REQUESTED** [review]

File: src/kubernetes_operator/iam_mapping.py
Comment (tbronsonCoveo) [reply src/kubernetes_operator/iam_mapping.py:75]: To explain further, the original operator overrides mapUsers with what it finds in the IamIdentityMappings, even if there were other entries unrelated to IamIdentityMappings (which didn't matter since there is nothing else there currently).
 
Transferring this logic to mapRoles would then also make it override everything including role mappings that are there and unrelated to the IamIdentityMappings.

Comment (drobitaille) [reply src/kubernetes_operator/iam_mapping.py:75]: These roles are there for a reason, we need access while UAM didn't pass yet. Why would we want to remove them?

Comment (JPLachance) [reply src/kubernetes_operator/iam_mapping.py:75]: Hi Dave,

These roles might be useful when UAM did not pass yet, but UAM is passing quite quickly and in my opinion, it is not an issue.

In addition to by-passing our user access management system, the way the roles mapping is done at the moment is making it impossible to retrace who did what in the cluster, because in the Kubernetes audit trail, we won't see "Dave did X", we see "kubectl-access-user-infra did x".

With strongDM, the audit trail is not an issue, because strongDM becomes the owner of that audit trail.

To recap, having a way to bypass our access management system that also breaks the required traceability is not ideal in my opinion.

I would:
- Add code to ignore the `system:node:{{EC2PrivateDNSName}}` user of the `mapRoles` list
- Flush all other entries that are not defined in a proper `IAMIdentityMapping` object

That being said, if you absolutely want to keep your UAM bypass, simply setup a `IAMIdentityMapping` outside of the `aws-auth` ConfigMap and our Operator will apply it.

So, what should we do? 😄

Comment (drobitaille) [reply src/kubernetes_operator/iam_mapping.py:75]: "That being said, if you absolutely want to keep your UAM bypass, simply setup a IAMIdentityMapping outside of the aws-auth ConfigMap and our Operator will apply it." 

This makes it possible for test cluster, etc. so ia have nothing to add. Its all good for me.

As for the actual "breaking" situation, i suppose it is feasable to do a progressive rollout that tolerate both for a while?

Comment (JPLachance) [reply src/kubernetes_operator/iam_mapping.py:75]: Tristen can do a PR in infra-kubernetes to add the IAMIdentityMapping you require.

On the first UAM pass, UAM will delete the IAMIdentityMapping to make the cluster compliant.

Then we can remove hardcoded access rights in the aws-auth CM, upgrade the Operator.

Good?

Comment (sbeginCoveo) [reply src/kubernetes_operator/iam_mapping.py:75]: est-ce que ca pourrait affecter le deploy_role que terraform utilise pour les provider k8s et helm?

Comment (JPLachance) [reply src/kubernetes_operator/iam_mapping.py:75]: La mécanique entre Terraform et Kubernetes m'est inconnu.

Est-ce que @jocgir peux nous en dire plus?

Comment (sbeginCoveo) [reply src/kubernetes_operator/iam_mapping.py:75]: En faite, la coupure sera au moment de la migration, donc fort probablement négligeable. votre plan semble etre solide :+1:

Comment (tbronsonCoveo) [reply src/kubernetes_operator/iam_mapping.py:75]: To make it work, there a lot of details to take into consideration. I am currently working on a plan to make sure everything runs smoothly. I'll post back next week when I've tested everything I could and I came up with a solid migration plan.

**Comment by tbronsonCoveo** [comment]
@JPLachance @sbeginCoveo @drobitaille The operator is good to deploy into dev. I made the corrections directly related to small code changes.

As stated before, the operator will not wipe existing map roles. I have tested locally with a replica of the existing configmap. 
**Plan**
1. Deploy updated operator image (https://hub.docker.com/r/coveo/aws-auth-operator) I am not sure how to go about that if someone could point me in the right direction.
2. Update operator image version in kubernetes/auth-operator.yaml with new image version
3. Merge this PR (btw I don't have the permissions, even with approvals)
4. Update deployment in dev us-east-1 of the operator (with the kubernetes/auth-operator.yaml)
5. Monitor for issues

If possible, I would like to have someone from infra on standby while I do the deployment.
Is that good with you guys?

**Comment by sbeginCoveo** [comment]
@tbronsonCoveo 
1. We release using [git tags](https://github.com/coveooss/aws_auth_eks_crd/tags)
4. [The deployment](https://github.com/coveo/infra-kubernetes/blob/master/infrastructure/terraform/coveo-eks/modules/auth_operator/deploy.tf) is part of infra-kubernetes and the version used is configured [here](https://github.com/coveo/infra-kubernetes/blob/master/infrastructure/built_docker_images.json#L76) So this will require a PR and deployment for this repo

**Review by JPLachance — APPROVED** [review]

**Review by sbeginCoveo — APPROVED** [review]

**Review by drobitaille — APPROVED** [review]

File: src/kubernetes_operator/iam_mapping.py
Comment (JPLachance) [inline src/kubernetes_operator/iam_mapping.py:33]: I really thing this should be configureable using an environment variable.

**Pull request merged**
